packages/cohort/src/replicator/services/statemap_queue_service.rs

```diff
@@ -0,0 +1,150 @@
+// $coverage:ignore-start
+
+use std::time::Duration;
+
+use futures::Future;
+use log::{error, info};
+use time::OffsetDateTime;
+use tokio::sync::mpsc;
+
+use crate::replicator::{
+    core::{StatemapInstallState, StatemapInstallationStatus, StatemapInstallerHashmap, StatemapItem},
+    models::StatemapInstallerQueue,
+};
+
+#[derive(Debug)]
+pub struct StatemapQueueServiceConfig {
+    pub queue_cleanup_frequency_ms: u64,
+    pub enable_stats: bool,
+}
+
+impl Default for StatemapQueueServiceConfig {
+    fn default() -> Self {
+        Self {
+            queue_cleanup_frequency_ms: 10_000,
+            enable_stats: true,
+        }
+    }
+}
+
+pub async fn statemap_queue_service(
+    mut statemaps_rx: mpsc::Receiver<Vec<StatemapItem>>,
+    mut statemap_installation_rx: mpsc::Receiver<StatemapInstallationStatus>,
+    installation_tx: mpsc::Sender<(u64, Vec<StatemapItem>)>,
+    // Get snapshot callback fn
+    get_snapshot_fn: impl Future<Output = Result<u64, String>>,
+    config: StatemapQueueServiceConfig,
+) -> Result<(), String> {
+    info!("Starting Installer Queue Service.... ");
+    let mut cleanup_interval = tokio::time::interval(Duration::from_millis(config.queue_cleanup_frequency_ms));
+
+    let mut installation_success_count = 0;
+    let mut installation_gaveup = 0;
+    let mut send_for_install_count = 0;
+    let mut first_install_start: i128 = 0; //
+    let mut last_install_end: i128 = 0; //  = OffsetDateTime::now_utc().unix_timestamp_nanos();
```

Comment:
A little mutable structure might be useful here.

Reply:
Agree, I had kept that for later and forgot about it. I will update it in the next PR when I move the replicator as its own package.